### PR TITLE
[LLK] fast reduce unpack for 32x32 tiles

### LINKED ISSUE #49779: Softmax perf regression from reduce_row algorithm change in #47311
#47311 removed the matmul_reduce path and now always uses the LLK reduce_tile path for REDUCE_ROW SUM/AVG. This introduced a performance regression in softmax and any op that relies on REDUCE_ROW.

The old matmul path called a single UNPACR instruction for the entire tile, unpacking all 4 faces contiguously in one shot. The current reduce path calls 4 separate UNPACR instructions. This actually slows down the math because it sets and clears dvalid 4 times instead of once, which introduces much higher latency in the pipeline.

### PR DESCRIPTION
**Pull request by markoradosavljevicTT**
### PR Category

Performance

### Summary

Closes #49779

This PR adds a full tile fast path for standard 32x32 tiles in both the unpack and math reduce kernels on Wormhole and Blackhole. For reduce_row SUM/AVG with 4 faces, the unpack now uses a single contiguous UNPACR per operand, and the math only has to clear dvalid once instead of 4 times per tile.

The SDPA perf test expected math utilization is lowered because the reduce is now faster, so the math spends less time waiting on dvalid. The overall L1 to L1 time is shorter, but the math utilization percentage drops because the math is no longer stalling doing nothing.

### Perf results

**Wormhole before:**

| formats.input_A | formats.input_B | formats.output | unpack_to_dest | dest_acc            | mathop                  | reduce_pool_type | tile_cnt | loop_factor | marker    | mean(L1_TO_L1) | mean(UNPACK_ISOLATE) | mean(MATH_ISOLATE) | mean(PACK_ISOLATE) | mean(L1_CONGESTION[UNPACK]) | mean(L1_CONGESTION[PACK]) | TEXT_SIZE(L1_TO_L1) | TEXT_SIZE(UNPACK_ISOLATE) | TEXT_SIZE(MATH_ISOLATE) | TEXT_SIZE(PACK_ISOLATE) |
| --------------- | --------------- | -------------- | -------------- | ------------------- | ----------------------- | ---------------- | -------- | ----------- | --------- | -------------- | -------------------- | ------------------ | ------------------ | --------------------------- | ------------------------- | ------------------- | ------------------------- | ----------------------- | ----------------------- |
| Float32         | Float32         | Float32        |          False | DestAccumulation.No | MathOperation.ReduceRow | ReducePool.Sum   |       16 |        True | INIT      |          572.0 |                437.0 |              254.0 |              571.0 |                       438.0 |                     563.0 |              12,601 |                     3,777 |                   3,434 |                   5,258 |
| Float32         | Float32         | Float32        |          False | DestAccumulation.No | MathOperation.ReduceRow | ReducePool.Sum   |       16 |        True | KERNEL    |        3,927.0 |              2,862.0 |            3,150.0 |            1,987.0 |                     3,151.0 |                   3,051.0 |              12,601 |                     3,777 |                   3,434 |                   5,258 |
| Float32         | Float32         | Float32        |          False | DestAccumulation.No | MathOperation.ReduceRow | ReducePool.Sum   |       16 |        True | TILE_LOOP |          207.5 |                140.5 |              168.7 |               77.1 |                       158.8 |                     144.2 |              12,601 |                     3,777 |                   3,434 |                   5,258 |

**Wormhole after:**

| formats.input_A | formats.input_B | formats.output | unpack_to_dest | dest_acc            | mathop                  | reduce_pool_type | tile_cnt | loop_factor | marker    | mean(L1_TO_L1) | mean(UNPACK_ISOLATE) | mean(MATH_ISOLATE) | mean(PACK_ISOLATE) | mean(L1_CONGESTION[UNPACK]) | mean(L1_CONGESTION[PACK]) | TEXT_SIZE(L1_TO_L1) | TEXT_SIZE(UNPACK_ISOLATE) | TEXT_SIZE(MATH_ISOLATE) | TEXT_SIZE(PACK_ISOLATE) |
| --------------- | --------------- | -------------- | -------------- | ------------------- | ----------------------- | ---------------- | -------- | ----------- | --------- | -------------- | -------------------- | ------------------ | ------------------ | --------------------------- | ------------------------- | ------------------- | ------------------------- | ----------------------- | ----------------------- |
| Float32         | Float32         | Float32        |          False | DestAccumulation.No | MathOperation.ReduceRow | ReducePool.Sum   |       16 |        True | INIT      |          575.0 |                440.0 |              232.0 |              562.0 |                       444.0 |                     567.0 |              12,809 |                     3,789 |                   3,630 |                   5,258 |
| Float32         | Float32         | Float32        |          False | DestAccumulation.No | MathOperation.ReduceRow | ReducePool.Sum   |       16 |        True | KERNEL    |        3,683.0 |              2,816.0 |            1,520.0 |            1,977.0 |                     3,191.0 |                   2,997.0 |              12,809 |                     3,789 |                   3,630 |                   5,258 |
| Float32         | Float32         | Float32        |          False | DestAccumulation.No | MathOperation.ReduceRow | ReducePool.Sum   |       16 |        True | TILE_LOOP |          191.9 |                137.6 |               68.2 |               77.2 |                       160.8 |                     140.6 |              12,809 |                     3,789 |                   3,630 |                   5,258 |

**Blackhole before:**

| formats.input_A | formats.input_B | formats.output | unpack_to_dest | dest_acc            | mathop                  | reduce_pool_type | tile_cnt | loop_factor | marker    | mean(L1_TO_L1) | mean(UNPACK_ISOLATE) | mean(MATH_ISOLATE) | mean(PACK_ISOLATE) | mean(L1_CONGESTION[UNPACK]) | mean(L1_CONGESTION[PACK]) | TEXT_SIZE(L1_TO_L1) | TEXT_SIZE(UNPACK_ISOLATE) | TEXT_SIZE(MATH_ISOLATE) | TEXT_SIZE(PACK_ISOLATE) |
| --------------- | --------------- | -------------- | -------------- | ------------------- | ----------------------- | ---------------- | -------- | ----------- | --------- | -------------- | -------------------- | ------------------ | ------------------ | --------------------------- | ------------------------- | ------------------- | ------------------------- | ----------------------- | ----------------------- |
| Float32         | Float32         | Float32        |          False | DestAccumulation.No | MathOperation.ReduceRow | ReducePool.Sum   |       16 |        True | INIT      |          484.0 |                461.0 |              209.0 |              482.0 |                       461.0 |                     473.0 |              11,721 |                     3,861 |                   3,326 |                   4,386 |
| Float32         | Float32         | Float32        |          False | DestAccumulation.No | MathOperation.ReduceRow | ReducePool.Sum   |       16 |        True | KERNEL    |        3,553.0 |              1,347.0 |            3,129.0 |            1,302.0 |                     1,345.0 |                   1,307.0 |              11,721 |                     3,861 |                   3,326 |                   4,386 |
| Float32         | Float32         | Float32        |          False | DestAccumulation.No | MathOperation.ReduceRow | ReducePool.Sum   |       16 |        True | TILE_LOOP |          184.0 |               46.1   |              173.1 |               41.4 |                        46.0 |                      42.3 |              11,721 |                     3,861 |                   3,326 |                   4,386 |

**Blackhole after:**

| formats.input_A | formats.input_B | formats.output | unpack_to_dest | dest_acc            | mathop                  | reduce_pool_type | tile_cnt | loop_factor | marker    | mean(L1_TO_L1) | mean(UNPACK_ISOLATE) | mean(MATH_ISOLATE) | mean(PACK_ISOLATE) | mean(L1_CONGESTION[UNPACK]) | mean(L1_CONGESTION[PACK]) | TEXT_SIZE(L1_TO_L1) | TEXT_SIZE(UNPACK_ISOLATE) | TEXT_SIZE(MATH_ISOLATE) | TEXT_SIZE(PACK_ISOLATE) |
| --------------- | --------------- | -------------- | -------------- | ------------------- | ----------------------- | ---------------- | -------- | ----------- | --------- | -------------- | -------------------- | ------------------ | ------------------ | --------------------------- | ------------------------- | ------------------- | ------------------------- | ----------------------- | ----------------------- |
| Float32         | Float32         | Float32        |          False | DestAccumulation.No | MathOperation.ReduceRow | ReducePool.Sum   |       16 |        True | INIT      |          487.0 |                468.0 |              196.0 |              484.0 |                       468.0 |                     474.0 |              11,929 |                     3,873 |                   3,522 |                   4,386 |
| Float32         | Float32         | Float32        |          False | DestAccumulation.No | MathOperation.ReduceRow | ReducePool.Sum   |       16 |        True | KERNEL    |        1,873.0 |              1,350.0 |            1,429.0 |            1,304.0 |                     1,351.0 |                   1,314.0 |              11,929 |                     3,873 |                   3,522 |                   4,386 |
| Float32         | Float32         | Float32        |          False | DestAccumulation.No | MathOperation.ReduceRow | ReducePool.Sum   |       16 |        True | TILE_LOOP |           78.7 |               45.8   |               67.1 |               41.4 |                        45.8 |                      42.7 |              11,929 |                     3,873 |                   3,522 |                   4,386 |

### CI Status
_Auto-generated on every push. Badges update live. Click a badge to filter runs by this branch._

- [](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:marko/reduce-full-tile-unpack)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/runtime-sanity-tests.yaml?query=branch:marko/reduce-full-tile-unpack)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-sanity-tests.yaml?query=branch:marko/reduce-full-tile-unpack)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:marko/reduce-full-tile-unpack)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/all-model-tests.yaml?query=branch:marko/reduce-full-tile-unpack)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:marko/reduce-full-tile-unpack)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:marko/reduce-full-tile-unpack)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:marko/reduce-full-tile-unpack)

### CODE REVIEW

File: tests/nightly/blackhole/sdpa/test_ring_joint_sdpa.py [hunk 1]
```diff
@@ -3492,7 +3493,7 @@ def test_ring_joint_attention_create_perf_table(model_name):
     RING_JOINT_PERF_CHECK_CONFIGS = [
         # (model_name, q_chunk_size, k_chunk_size, ring_size, expected_util, margin)
         # 4-device ring (QuietBox, sp=4 tp=1)
-        ("wan2_2_1xGLX", 288, 512, 4, 68.9, RING_JOINT_PERF_MARGIN),
+        ("wan2_2_1xGLX", 288, 512, 4, 68.3, RING_JOINT_PERF_MARGIN),
```
Comment (skrsticTT) [inline tests/nightly/blackhole/sdpa/test_ring_joint_sdpa.py]: why we have this perf drop? it is actually significant, is actual main per number lower than 68.9, and your change just put it bellow lower threshold, or this is just regression from this pr?

File: tests/nightly/blackhole/sdpa/test_ring_joint_sdpa.py [hunk 2]
```diff
@@ -29,11 +29,10 @@
 
 import pytest
 import torch
-from loguru import logger
-from ttnn.operations.ccl import Topology
-
 import ttnn
+from loguru import logger
 from models.common.utility_functions import skip_with_llk_assert, skip_with_watcher
+from ttnn.operations.ccl import Topology
```
Comment (skrsticTT) [inline tests/nightly/blackhole/sdpa/test_ring_joint_sdpa.py:33]: seems like nothing changed here?

File: tests/nightly/blackhole/sdpa/test_ring_joint_sdpa.py [hunk 3]
```diff
@@ -329,7 +328,9 @@ def scaled_model_heads_for_mesh(model: ModelConfig, mesh_config: MeshConfig) ->
 from tests.nightly.sdpa_perf_utils import (
     ARCH_CONSTANTS,
 )
-from tests.nightly.sdpa_perf_utils import compute_math_utilization as compute_ring_joint_utilization
+from tests.nightly.sdpa_perf_utils import (
+    compute_math_utilization as compute_ring_joint_utilization,
```
Comment (skrsticTT) [inline tests/nightly/blackhole/sdpa/test_ring_joint_sdpa.py]: same here

File: tests/nightly/blackhole/sdpa/test_ring_joint_sdpa.py (hunk 1 above)
Comment (markoradosavljevicTT) [reply tests/nightly/blackhole/sdpa/test_ring_joint_sdpa.py]: This isn't a performance drop , it's a utilization drop. In the performance analyses I posted in the description, you can see that the math is executed in roughly half the number of cycles.

File: tests/nightly/blackhole/sdpa/test_ring_joint_sdpa.py (hunk 2 above)
Comment (markoradosavljevicTT) [reply tests/nightly/blackhole/sdpa/test_ring_joint_sdpa.py:33]: This is just a linter change

File: tests/nightly/blackhole/sdpa/test_ring_joint_sdpa.py (hunk 1 above)
Comment (fvranicTT) [reply tests/nightly/blackhole/sdpa/test_ring_joint_sdpa.py]: @markoradosavljevicTT let's briefly explain those Perf tables you put in the PR description. It's hard to understand what you told me yesterday just by looking at the tables in isolation. Thanks!

Comment (markoradosavljevicTT) [reply tests/nightly/blackhole/sdpa/test_ring_joint_sdpa.py]: In the perf tables I posted, it's important to pay attention to the `L1_TO_L1`, `UNPACK_ISOLATE`, and `MATH_ISOLATE` columns. Unpack remained just as fast as before, and the complete L1 to L1 is bottlenecked by unpack. The difference is that now math doesn't have to wait 4 times for data in the src registers and doesn't lose time waiting for data between faces.

**Comment by markoradosavljevicTT** [comment]
Perf regression tests are passing: https://github.com/tenstorrent/tt-metal/actions/runs/29236357000

File: tests/nightly/blackhole/sdpa/test_ring_joint_sdpa.py (hunk 1 above)
Comment (skrsticTT) [reply tests/nightly/blackhole/sdpa/test_ring_joint_sdpa.py]: > This isn't a performance drop , it's a utilization drop. 

math utilization is our perf metric here, we calculate how our fpu is busy compared to theoretical number, it is basically measured time divided by time if fpu was busy all the time
in the simple words, if op is faster -> math util is higher

Comment (skrsticTT) [reply tests/nightly/blackhole/sdpa/test_ring_joint_sdpa.py]: we have lower and upper band, so maybe regression came earlier and you just put it bellow threshold. can you check diff between your change and main?

Comment (markoradosavljevicTT) [reply tests/nightly/blackhole/sdpa/test_ring_joint_sdpa.py]: Here are the results from main and from my branch, and on main the result is not what we expect. On my branch it's only 0.01% lower.
```
main:

2026-07-14 13:11:29.505 | INFO     | tests.nightly.blackhole.sdpa.test_ring_joint_sdpa:test_ring_joint_attention_perf_check:3548 - Ring joint SDPA perf check wan2_2_1xGLX-q288-k512: duration=7.895 ms, math_util=68.49% (expected 68.90%, band [68.21, 69.59]), profiler_records=4
2026-07-14 13:12:22.759 | INFO     | tests.nightly.blackhole.sdpa.test_ring_joint_sdpa:test_ring_joint_attention_perf_check:3548 - Ring joint SDPA perf check wan2_2_1xGLX-q288-k512: duration=7.896 ms, math_util=68.48% (expected 68.90%, band [68.21, 69.59]), profiler_records=4
2026-07-14 13:12:53.722 | INFO     | tests.nightly.blackhole.sdpa.test_ring_joint_sdpa:test_ring_joint_attention_perf_check:3548 - Ring joint SDPA perf check wan2_2_1xGLX-q288-k512: duration=7.895 ms, math_util=68.49% (expected 68.90%, band [68.21, 69.59]), profiler_records=4
2026-07-14 13:13:22.487 | INFO     | tests.nightly.blackhole.sdpa.test_ring_joint_sdpa:test_ring_joint_attention_perf_check:3548 - Ring joint SDPA perf check wan2_2_1xGLX-q288-k512: duration=7.895 ms, math_util=68.49% (expected 68.90%, band [68.21, 69.59]), profiler_records=4

marko/reduce-full-tile-unpack:

2026-07-14 13:23:51.168 | INFO     | tests.nightly.blackhole.sdpa.test_ring_joint_sdpa:test_ring_joint_attention_perf_check:3549 - Ring joint SDPA perf check wan2_2_1xGLX-q288-k512: duration=7.897 ms, math_util=68.47% (expected 68.30%, band [67.62, 68.98]), profiler_records=4
2026-07-14 13:24:27.949 | INFO     | tests.nightly.blackhole.sdpa.test_ring_joint_sdpa:test_ring_joint_attention_perf_check:3549 - Ring joint SDPA perf check wan2_2_1xGLX-q288-k512: duration=7.897 ms, math_util=68.48% (expected 68.30%, band [67.62, 68.98]), profiler_records=4
2026-07-14 13:24:51.953 | INFO     | tests.nightly.blackhole.sdpa.test_ring_joint_sdpa:test_ring_joint_attention_perf_check:3549 - Ring joint SDPA perf check wan2_2_1xGLX-q288-k512: duration=7.899 ms, math_util=68.46% (expected 68.30%, band [67.62, 68.98]), profiler_records=4
2026-07-14 13:25:34.812 | INFO     | tests.nightly.blackhole.sdpa.test_ring_joint_sdpa:test_ring_joint_attention_perf_check:3549 - Ring joint SDPA perf check wan2_2_1xGLX-q288-k512: duration=7.896 ms, math_util=68.48% (expected 68.30%, band [67.62, 68.98]), profiler_records=4

```

Comment (markoradosavljevicTT) [reply tests/nightly/blackhole/sdpa/test_ring_joint_sdpa.py]: @skrsticTT my change shouldn't affect this test at all because it doesn't use this kernel. It looks like some earlier change on main made this test flaky. Here is the sanity test that failed for me: https://github.com/tenstorrent/tt-metal/actions/runs/29258362605/job/86856975660

File: tests/nightly/blackhole/sdpa/test_ring_joint_sdpa.py [hunk 4]
```diff
@@ -29,11 +29,10 @@
 
 import pytest
 import torch
-from loguru import logger
-from ttnn.operations.ccl import Topology
-
 import ttnn
+from loguru import logger
```
Comment (fvranicTT) [inline tests/nightly/blackhole/sdpa/test_ring_joint_sdpa.py:32]: Can we revert these imports to reduce the noise in this file? Effectively, only line 3496 is important here. Thanks.

File: tests/nightly/blackhole/sdpa/test_ring_joint_sdpa.py (hunk 1 above)
Comment (skrsticTT) [reply tests/nightly/blackhole/sdpa/test_ring_joint_sdpa.py]: ok, can we maybe then remove this change because it is unrelated to your changes, and open an issue (you can assign it on me) that that perf test is flaky on main?

**Review by nvelickovicTT — APPROVED** [review]
